src/wasm/wat-parser.cpp

```diff
@@ -0,0 +1,1809 @@
+
+  RETURN_OR_OK(params);
+}
+
+// result   ::= '(' 'result' t*:valtype ')' => [t*]
+template<typename Ctx>
+Result<typename Ctx::Results> result(Ctx& ctx, ParseInput& in) {
+  if (!in.takeSExprStart("result"sv)) {
+    return {};
+  }
+
+  std::vector<Type> results;
+  while (!in.takeRParen()) {
+    auto type = valtype(ctx, in);
+    if (!type) {
+      return {};
+    }
+    if constexpr (!parsingDecls<Ctx>) {
+      results.push_back(*type);
+    }
+  }
+
+  RETURN_OR_OK(results);
+}
+
+// functype ::= '(' 'func' t1*:vec(param) t2*:vec(result) ')' => [t1*] -> [t2*]
+template<typename Ctx>
+Result<typename Ctx::FuncType> functype(Ctx& ctx, ParseInput& in) {
+  if (!in.takeSExprStart("func"sv)) {
+    return {};
+  }
+
+  std::vector<Type> params, results;
+  while (auto newParams = param(ctx, in)) {
+    if constexpr (!parsingDecls<Ctx>) {
+      for (auto& p : *newParams) {
+        params.push_back(p.type);
+      }
+    }
+  }
+
+  while (auto newResults = result(ctx, in)) {
+    if constexpr (!parsingDecls<Ctx>) {
+      results.insert(results.end(), newResults->begin(), newResults->end());
+    }
+  }
+
+  if (!in.takeRParen()) {
+    return {};
+  }
+
+  RETURN_OR_OK(Signature(Type(params), Type(results)));
+}
+
+// limits ::= n:u32       => { min n, max _ }
+//          | n:u32 m:u32 => { min n, max m }
+Result<Limits> limits(ParseInput& in) {
+  auto min = in.takeU32();
+  if (!min) {
+    return {};
+  }
+
+  return {{*min, in.takeU32()}};
+}
+
+// memtype ::= lim:limits => lim
+Result<Limits> memtype(ParseInput& in) { return limits(in); }
+
+// tabletype ::= lim:limits et:reftype => lim et
+template<typename Ctx>
+Result<typename Ctx::TableType> tabletype(Ctx& ctx, ParseInput& in) {
+  auto lim = limits(in);
+  if (!lim) {
+    return {};
+  }
+
+  auto type = reftype(ctx, in);
+  if (!type) {
+    return {};
+  }
+
+  if constexpr (parsingDecls<Ctx>) {
+    return Ok{};
+  } else if constexpr (parsingDefs<Ctx>) {
+    return {{*lim, *type}};
+  }
+}
+
+// globaltype ::= t:valtype               => const t
+//              | '(' 'mut' t:valtype ')' => var t
+template<typename Ctx>
+Result<typename Ctx::GlobalType> globaltype(Ctx& ctx, ParseInput& in) {
+  // t:valtype
+  if (auto type = valtype(ctx, in)) {
+    if constexpr (parsingDecls<Ctx>) {
+      return Ok{};
+    } else if constexpr (parsingDefs<Ctx>) {
+      return {{Immutable, *type}};
+    }
+  }
+
+  // '(' 'mut' t:valtype ')'
+  if (!in.takeSExprStart("mut"sv)) {
+    return {};
+  }
+  auto type = valtype(ctx, in);
+  if (!type) {
+    return {};
+  }
+  if (!in.takeRParen()) {
+    return {};
+  }
+
+  if constexpr (parsingDecls<Ctx>) {
+    return Ok{};
+  } else if constexpr (parsingDefs<Ctx>) {
+    return {{Mutable, *type}};
+  }
+}
+
+// TODO
+template<typename Ctx> Result<typename Ctx::Instrs> instrs(Ctx&, ParseInput&) {
+  return {};
+}
+
+// expr ::= (in:instr)* => in* end
+template<typename Ctx> Result<typename Ctx::Instrs> expr(Ctx&, ParseInput&) {
+  // TODO
```

Comment:
Is this where `scripts/gen-s-parser.py` integration will be?

Reply:
Yes, we will use that generated code to generate function calls for each individual instruction given the keyword (i.e. instruction name) in this function or a child.